File: backend/internal/base/datatype/userrole/userrole.go
// Package userrole defines the individual end-user roles used for local RBAC
// (checked via UserRoles.HasRoles in command/query handlers). This is a
// separate authorization layer from peer-scoped task ownership: a contract's
// Responsible.Approvers/Reviewers/Negotiators are peer DIDs (which DCS
// instance is responsible for a task), while UserRole governs which
// individual, locally authenticated user may act on behalf of that peer.
package userrole

import (
	"fmt"
)

type UserRole string

const (
	TemplateCreator  UserRole = "Template Creator"
	TemplateReviewer UserRole = "Template Reviewer"
	TemplateApprover UserRole = "Template Approver"
	TemplateManager  UserRole = "Template Manager"

	ContractCreator    UserRole = "Contract Creator"
	ContractReviewer   UserRole = "Contract Reviewer"
	ContractApprover   UserRole = "Contract Approver"
	ContractManager    UserRole = "Contract Manager"
	ContractNegotiator UserRole = "Contract Negotiator"
	ContractSigner     UserRole = "Contract Signer"
	ContractObserver   UserRole = "Contract Observer"

	ArchiveManager      UserRole = "Archive Manager"
	Auditor             UserRole = "Auditor"
	SystemAdministrator UserRole = "Sys. Administrator"
	ComplianceOfficer   UserRole = "Compliance Officer"
	IntegrationManager  UserRole = "Integration Manager"

	ProcessOrchestrator UserRole = "Process Orchestrator"
	Validator           UserRole = "Validator"

	SystemContractCreator  UserRole = "Sys. Contract Creator"
	SystemContractReviewer UserRole = "Sys. Contract Reviewer"
	SystemContractApprover UserRole = "Sys. Contract Approver"
	SystemContractManager  UserRole = "Sys. Contract Manager"
	SystemContractSigner   UserRole = "Sys. Contract Signer"
	ContractTargetSystem   UserRole = "Contract Target System"
)

func NewUserRole(s string) (UserRole, error) {
	ts := UserRole(s)
	if !ts.IsValid() {
		return "", fmt.Errorf("invalid user role state: %s", s)
	}
	return ts, nil
}

// IsValid checks if the UserRole is a valid role
func (r UserRole) IsValid() bool {
	switch r {
	case TemplateCreator, TemplateReviewer, TemplateApprover, TemplateManager,
		ContractCreator, ContractReviewer, ContractApprover, ContractManager,
		ContractNegotiator, ContractSigner, ContractObserver,
		ArchiveManager, Auditor, SystemAdministrator, ComplianceOfficer, IntegrationManager,
		ProcessOrchestrator, Validator,
		SystemContractCreator, SystemContractReviewer, SystemContractApprover,
		SystemContractManager, SystemContractSigner, ContractTargetSystem:
		return true
	}
	return false
}

// String returns the string representation of the UserRole
func (r UserRole) String() string {
	return string(r)
}

// IsSystemRole returns true if the role is a system/automated role
func (r UserRole) IsSystemRole() bool {
	switch r {
	case SystemContractCreator, SystemContractReviewer, SystemContractApprover,
		SystemContractManager, SystemContractSigner, ContractTargetSystem:
		return true
	}
	return false
}

// IsHumanRole returns true if the role is a human user role
func (r UserRole) IsHumanRole() bool {
	return r.IsValid() && !r.IsSystemRole()
}

type UserRoles []UserRole

func (r UserRoles) HasRoles(requiredRoles ...UserRole) bool {
	for _, requiredRole := range requiredRoles {
		for _, role := range r {
			if role == requiredRole {
				return true
			}
		}
	}
	return false
}

File: frontend/ClientApp/src/types/user-role.ts
export type UserRole =
  | 'TEMPLATE_CREATOR'
  | 'TEMPLATE_REVIEWER'
  | 'TEMPLATE_APPROVER'
  | 'TEMPLATE_MANAGER'
  | 'CONTRACT_CREATOR'
  | 'CONTRACT_REVIEWER'
  | 'CONTRACT_APPROVER'
  | 'CONTRACT_MANAGER'
  | 'CONTRACT_NEGOTIATOR'
  | 'CONTRACT_SIGNER'
  | 'CONTRACT_OBSERVER'
  | 'ARCHIVE_MANAGER'
  | 'AUDITOR'
  | 'SYSTEM_ADMINISTRATOR'
  | 'COMPLIANCE_OFFICER'
  | 'INTEGRATION_MANAGER'
  | 'PROCESS_ORCHESTRATOR'
  | 'VALIDATOR'

/** Maps access-token role claim labels to UserRole ids. */
const ROLE_LABEL_TO_USER_ROLE: Record<string, UserRole> = {
  'Template Creator': 'TEMPLATE_CREATOR',
  'Template Reviewer': 'TEMPLATE_REVIEWER',
  'Template Approver': 'TEMPLATE_APPROVER',
  'Template Manager': 'TEMPLATE_MANAGER',
  'Contract Creator': 'CONTRACT_CREATOR',
  'Contract Reviewer': 'CONTRACT_REVIEWER',
  'Contract Approver': 'CONTRACT_APPROVER',
  'Contract Manager': 'CONTRACT_MANAGER',
  'Contract Negotiator': 'CONTRACT_NEGOTIATOR',
  'Contract Signer': 'CONTRACT_SIGNER',
  'Contract Observer': 'CONTRACT_OBSERVER',
  'Archive Manager': 'ARCHIVE_MANAGER',
  Auditor: 'AUDITOR',
  'Sys. Administrator': 'SYSTEM_ADMINISTRATOR',
  'Compliance Officer': 'COMPLIANCE_OFFICER',
  'Integration Manager': 'INTEGRATION_MANAGER',
  'Process Orchestrator': 'PROCESS_ORCHESTRATOR',
  Validator: 'VALIDATOR',
}

/** Reads roles from a JWT payload. */
export function rolesFromJwtPayload(payload: Record<string, unknown> | null | undefined): unknown {
  if (!payload) return []
  if (Array.isArray(payload.roles)) return payload.roles
  const ext = payload.ext
  if (ext && typeof ext === 'object' && Array.isArray((ext as Record<string, unknown>).roles)) {
    return (ext as Record<string, unknown>).roles
  }
  return []
}

export function mapRoleLabelsToUserRoles(roles: unknown): UserRole[] {
  if (!Array.isArray(roles)) return []
  const mapped: UserRole[] = []
  for (const r of roles) {
    if (typeof r !== 'string') continue
    const role = ROLE_LABEL_TO_USER_ROLE[r]
    if (role) mapped.push(role)
  }
  return mapped
}
